core/multigrid/pgm.cpp

```diff
@@ -227,10 +237,238 @@ void Pgm<ValueType, IndexType>::generate()
+        host_send_buffer.resize_and_reset(total_send_size);
+        exec->get_master()->copy_from(exec, total_send_size,
+                                      send_agg.get_data(),
+                                      host_send_buffer.get_data());
+    }
+    auto type = experimental::mpi::type_impl<IndexType>::get_type();
+
+    const auto send_ptr = use_host_buffer ? host_send_buffer.get_const_data()
+                                          : send_agg.get_const_data();
+    auto recv_ptr = use_host_buffer ? host_recv_buffer.get_data()
+                                    : non_local_agg.get_data();
+    exec->synchronize();
+    comm.all_to_all_v(use_host_buffer ? exec->get_master() : exec, send_ptr,
+                      send_sizes.data(), send_offsets.data(), type, recv_ptr,
+                      recv_sizes.data(), recv_offsets.data(), type);
+    if (use_host_buffer) {
+        exec->copy_from(exec->get_master(), total_recv_size, recv_ptr,
+                        non_local_agg.get_data());
+    }
+    return;
+}
+
+
+template <typename IndexType>
+struct larger_index {
+    using type = IndexType;
+};
+
+template <>
+struct larger_index<gko::int32> {
+    using type = gko::int64;
+};
 
-    this->set_multigrid_level(prolong_row_gather, coarse_matrix,
-                              restrict_sparsity);
+
+template <typename ValueType, typename IndexType>
+void Pgm<ValueType, IndexType>::generate()
+{
+    using csr_type = matrix::Csr<ValueType, IndexType>;
+#if GINKGO_BUILD_MPI
+    if (auto matrix = std::dynamic_pointer_cast<
+            const experimental::distributed::MatrixBase<IndexType>>(
+            system_matrix_)) {
+        // only work for the square local matrix
+        auto exec = gko::as<LinOp>(matrix)->get_executor();
+        auto comm = gko::as<experimental::distributed::DistributedBase>(matrix)
+                        ->get_communicator();
+        auto num_rank = comm.size();
+        // Only support csr matrix currently.
+        auto pgm_local_op = std::dynamic_pointer_cast<const csr_type>(
+            matrix->get_local_matrix());
+        // If system matrix is not csr or need sorting, generate the csr.
+        if (!parameters_.skip_sorting || !pgm_local_op) {
+            pgm_local_op = convert_to_with_sorting<csr_type>(
+                exec, matrix->get_local_matrix(), parameters_.skip_sorting);
+        }
+        auto result = this->generate_local(pgm_local_op);
+        auto non_local_matrix = matrix->get_non_local_matrix();
+        auto non_local_size = non_local_matrix->get_size()[1];
+        array<IndexType> non_local_agg(exec, non_local_size);
+        // prolong, restrict, only needs the local infomation
+        // get agg information (prolong_row_gather row idx)
+        communicate(matrix, agg_, non_local_agg);
+        // generate non_local_col_map
+        non_local_agg.set_executor(exec->get_master());
+        array<IndexType> non_local_col_map(exec->get_master(), non_local_size);
```

Comment:
I think computing the `non_local_col_map` here can be merged with a similar part in the `Matrix::read_distributed`. In both cases, what we are doing is, for an index sequence that is segmented by keys (the target rank), we map each index into the interval `[0, U)`, where `U` is the number of unique indices in that segment. I will extract the relevant parts. That should simplify the following code a lot.

Reply:
The branch `compress-indices` contains the relevant part of `read_distributed` extracted. Using the part_id as keys and the non_local_agg as indices should give you a mapping from the fine non-local columns to the coarse non-local columns. Since the output will also contain the part_ids for the coarse unique columns, it should also simplify computing the gather_idxs, and recv_sizes. I would guess that both can be computed nearly identically to the approach in `read_distibuted`